700 scan create info box

spaudanjo opened the pull request:
Closes https://trello.com/c/np5Z0eLP/700-make-a-box-on-qr-code-scan-follow-up-createboxtsx-react-flask-graphql-medium-xlarge


### Next steps / ideas for follow up tickets: 

* reviewing the resolver logic (including adding tests - the resolvers added in this PR were sketched out mainly to enable frontend development) - at the moment there is at least one bug: after a QR code is scanned and a box is created for it, the same QR code scanning doesn't lead to the details page of the box just created
  * this might probably be obsolete (or done together with) as soon as Philipp looks into redesigning the overall resolvers/schema implementation more profoundly as discussed today - then the frontend of course needs to adapt to the new/updated schemma of course
  * in general @pylipp: please don't let distract your resolver implementation work from the queries currently used in the frontend - they were added based on what was already available in the BE and based on adding additional resolvers in a fast/pragmatic manner to focus on FE dev/onboarding; also feel free to heavily adapt / remove the resolvers I added of course (if some FE logic would break because of that, I don't think this is bad since we could focus on adapting the FE in the next hackathon to the updated API) 

* tests for box-info/details page

* cleaning up UI of the scan box and box detail page (consistent usage of Soheimas ui library package) 

* correct implementation of the size selection / dropdown (related to schema discussion) 

* general review of the current frontend codebase and UI and UX (there are some bugs/funny behaviours atm)

pylipp commented on flask/boxwise_flask/graph_ql/query_defs.py:
```diff
@@ -10,9 +10,14 @@
+        qrCode(id: Int!): QrCode
         qrExists(qr_code: String): Boolean
-        qrBoxExists(qr_code: String): Boolean
+        boxIdByQrCode(qr_code: String): Int
```
I'd remove this query since one can write ```
query { qrCode (id: ID) { box { box_id } } }
```
instead

spaudanjo replied:
Yes, that would be perfect - but is this already implemented on master in the backend? And I think instead of id we need to pass in the qr_code to the qrCode() call, right?

pylipp replied:
It's part of the #59 PR, and about to come to master. Yep, has to be qr_code as an argument!

pylipp replied:
FYI the latest commit removes the boxIdByQrCode query.

pylipp commented on flask/boxwise_flask/graph_ql/query_defs.py:
```diff
@@ -10,9 +10,14 @@
         allUsers: [User]
         user(email: String): User
-        box(qr_code: String): Box
+        box(id: String): Box
```
Can we make the argument types more consistent? For box, location, qrCode, product it's a mix of String and Int. Also I remember a schema proposal from @aerinsol where they were of type ID.

spaudanjo replied:
Totally agree! I would suggest to postpone these important consistency adjustments and cleanups to a several work package / PR(s) though. 
As a reminder, the backend changes here in this PR are for sake of pragmatism for enabling the FE progress.

spaudanjo replied:
A good workflow in this early stage where the API / basic BE capabilities are still sketched out and FE devs are not working on a daily/weekly basis might be to ignore that BE/schema changes might break the current FE implementation. I will then follow up with the FE in the next hackathon to adapt the queries. Since FE tests are mocked, the pipeline wouldn't fail for FE and you wouldn't be blocked.

pylipp replied:
Alright I'll take that literally ;) and try to integrate the most recent schema here already.

pylipp reviewed (changes requested):
Excited to see how this looks like in a demo!
I'd like to discuss the query definitions a little before merging. Overall I suggest we strive for adhering to a single proposed schema for types and queries. I think there'll be a couple of merge conflicts with #59 but that's probably inevitable when a project is in an early state.

pylipp commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Make a Box on QR Code Scan follow-up - CreateBox.tsx #react #flask #graphql #medium #xlarge](https://trello.com/c/np5Z0eLP/700-make-a-box-on-qr-code-scan-follow-up-createboxtsx-react-flask-graphql-medium-xlarge)

pylipp commented:
When I run `docker-compose up` the start the entire system, the react container fails with
```
./src/Auth0ProviderWithHistory.tsx
Cannot find module: '@auth0/auth0-react'. Make sure this package is installed.
```
@spaudanjo Can you check if that's the same on your end?

spaudanjo commented:
> auth0/auth0-react

I start the react app without docker (I commented the react section out locally in the docker-compose file). So I also run yarn locally and there the installation succeeds and I can start the app.

spaudanjo commented:
I'm actually considering to remove the resolver changes from this PR (also because the pipeline is blocked atm because the resolvers don't have tests => reduction of test coverage) and wait for the more comprehensive resolver PR to come. What do you think @pylipp

pylipp commented:
> > auth0/auth0-react
> 
> @pylipp
> 
> I start the react app without docker (I commented the react section out locally in the docker-compose file). So I also run yarn locally and there the installation succeeds and I can start the app.

I see :D is there a way to have the Docker image updated? I don't have the JS stack on my machine^^

Re the resolvers they should have tests now (see my latest commits).

spaudanjo commented:
> > > auth0/auth0-react
> > 
> > 
> > @pylipp
> > I start the react app without docker (I commented the react section out locally in the docker-compose file). So I also run yarn locally and there the installation succeeds and I can start the app.
> 
> I see :D is there a way to have the Docker image updated? I don't have the JS stack on my machine^^
> 
> Re the resolvers they should have tests now (see my latest commits).

I will try to do it - I initially had issues with the docker-compose because of Mac M1 incompatibilities. Maybe this can be fixed now.

aerinsol commented:
Blog from Docker about M1 compatibility:
https://docs.docker.com/desktop/mac/apple-silicon/

Agree with Philipp that the config should be updated in Docker.

On Sun, Sep 19, 2021, 8:49 PM Daniel Spaude ***@***.***>
wrote:

> auth0/auth0-react
>
> @pylipp <https://github.com/pylipp>
> I start the react app without docker (I commented the react section out
> locally in the docker-compose file). So I also run yarn locally and there
> the installation succeeds and I can start the app.
>
> I see :D is there a way to have the Docker image updated? I don't have the
> JS stack on my machine^^
>
> Re the resolvers they should have tests now (see my latest commits).
>
> I will try to do it - I initially had issues with the docker-compose
> because of Mac M1 incompatibilities. Maybe this can be fixed now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/boxwise/boxtribute/pull/62#issuecomment-922468751>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABP2BFC7W2YO4DISXPLJMPLUCXL6DANCNFSM5EJNC6ZA>
> .
>

spaudanjo commented:
> Blog from Docker about M1 compatibility: https://docs.docker.com/desktop/mac/apple-silicon/ Agree with Philipp that the config should be updated in Docker.
> […](#)
> On Sun, Sep 19, 2021, 8:49 PM Daniel Spaude ***@***.***> wrote: auth0/auth0-react @pylipp <https://github.com/pylipp> I start the react app without docker (I commented the react section out locally in the docker-compose file). So I also run yarn locally and there the installation succeeds and I can start the app. I see :D is there a way to have the Docker image updated? I don't have the JS stack on my machine^^ Re the resolvers they should have tests now (see my latest commits). I will try to do it - I initially had issues with the docker-compose because of Mac M1 incompatibilities. Maybe this can be fixed now. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#62 (comment)](https://github.com/boxwise/boxtribute/pull/62#issuecomment-922468751)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABP2BFC7W2YO4DISXPLJMPLUCXL6DANCNFSM5EJNC6ZA> .

I setup the project from scratch on my old Intel Mac (also freshly installed Docker there) and the react docker container fails there as well. Looks like this deserve its own (hopefully small) task, like `Fixing React Docker setup`

vahidbazzaz commented:
@spaudanjo I had the same problem before, but it was caused by a memory limit on the Docker. Would you please check if you have given at least 4 GB to the Docker, the default setting is set to 2 GB when freshly installed. Hope this helps.

pylipp commented:
> > Blog from Docker about M1 compatibility: https://docs.docker.com/desktop/mac/apple-silicon/ Agree with Philipp that the config should be updated in Docker.
> > […](#)
> > On Sun, Sep 19, 2021, 8:49 PM Daniel Spaude _**@**_.***> wrote: auth0/auth0-react @pylipp https://github.com/pylipp I start the react app without docker (I commented the react section out locally in the docker-compose file). So I also run yarn locally and there the installation succeeds and I can start the app. I see :D is there a way to have the Docker image updated? I don't have the JS stack on my machine^^ Re the resolvers they should have tests now (see my latest commits). I will try to do it - I initially had issues with the docker-compose because of Mac M1 incompatibilities. Maybe this can be fixed now. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#62 (comment)](https://github.com/boxwise/boxtribute/pull/62#issuecomment-922468751)>, or unsubscribe https://github.com/notifications/unsubscribe-auth/ABP2BFC7W2YO4DISXPLJMPLUCXL6DANCNFSM5EJNC6ZA .
> 
> I setup the project from scratch on my old Intel Mac (also freshly installed Docker there) and the react docker container fails there as well. Looks like this deserve its own (hopefully small) task, like `Fixing React Docker setup`

I ran `docker build .` in the `react/` directory and received
```
Step 4/5 : RUN yarn install
 ---> Running in f481c77fe8c3
yarn install v1.22.5
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@2.1.2: The platform "linux" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.3.2: The platform "linux" is incompatible with this module.
info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.9: The platform "linux" is incompatible with this module.
info "fsevents@1.2.9" is an optional dependency and failed compatibility check. Excluding it from installation.
error jest-diff@27.1.1: The engine "node" is incompatible with this module. Expected version "^10.13.0 || ^12.13.0 || ^14.15.0 || >=15.0.0". Got "14.9.0"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
The command '/bin/sh -c yarn install' returned a non-zero code: 1
```
Maybe that's helpful, too.

pylipp commented:
> Blog from Docker about M1 compatibility: https://docs.docker.com/desktop/mac/apple-silicon/ Agree with Philipp that the config should be updated in Docker.
> […](#)
> On Sun, Sep 19, 2021, 8:49 PM Daniel Spaude ***@***.***> wrote: auth0/auth0-react @pylipp <https://github.com/pylipp> I start the react app without docker (I commented the react section out locally in the docker-compose file). So I also run yarn locally and there the installation succeeds and I can start the app. I see :D is there a way to have the Docker image updated? I don't have the JS stack on my machine^^ Re the resolvers they should have tests now (see my latest commits). I will try to do it - I initially had issues with the docker-compose because of Mac M1 incompatibilities. Maybe this can be fixed now. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#62 (comment)](https://github.com/boxwise/boxtribute/pull/62#issuecomment-922468751)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABP2BFC7W2YO4DISXPLJMPLUCXL6DANCNFSM5EJNC6ZA> .

Should fixing the Docker setup happen here, or do you want to have a separate task in the backlog @aerinsol ?

aerinsol commented:
> Should fixing the Docker setup happen here, or do you want to have a separate task in the backlog @aerinsol ?

Let's have a new task. Will link it here shortly.

aerinsol commented:
https://trello.com/c/BeRTxAji @spaudanjo @pylipp

spaudanjo commented:
general note: sorry folks that i'm not as responsive - outside of hackathons I only have - if at all - some minutes after work during the week :-/

spaudanjo commented:
@vahidbazzaz the memory increase worked - thanks! :D 
I just pushed some changes - with these for me at least locally the full docker-compose stack works now (including react) :) 
maybe @pylipp you could try once more?

spaudanjo commented:
next step for me would be to adapt the app to the new/changed queries @pylipp wrote (thanks for that and for pushing them directly here to this PR) :D

spaudanjo commented:
btw a proposal @HaGuesto @aerinsol @pylipp: 
Could we change the allowed merge strategy in github for PRs to `Rebase` (maybe also `Squash` - but I think rebase is the more critical one)? A history full of merge commits can sometimes be really annoying...

aerinsol commented:
Squash is default merge strategy now. Aren't you not supposed to rebase public repositories master? Keep in mind we use trunk based development, not gitflow

spaudanjo commented:
> Squash is default merge strategy now. Aren't you not supposed to rebase public repositories master? Keep in mind we use trunk based development, not gitflow

I rebase locally in general if I work alone on a branch. 
If we rebase locally we need to be careful if we are in a situations like this atm were >1 people push to the same branch (which is in fact a temporary exception of trunk based). 
So good point actually - @pylipp if it doesn't create coordinative issues for you, could you then in fact do your changes (as well) on a separate branch / stop pushing to this one? Then I will also rebase this branch on master now

pylipp commented:
Hej Daniel, all good! I moved a couple of messages to Slack to keep the postings here about PR-related stuff. I won't push to this branch anymore =) 
* For frontend coverage, I don't really have an opinion. For BE coverage I try to be strict and to cover whatever new production code is implemented
* rebasing the current branch on master could become tricky now that #59 was merged...maybe now it is a good idea to remove the resolver implementations and their tests again
* I still can't run the react service. With `docker-compose up react` I have the same error about `auth0-react` missing. When I run `docker build react/` before I receive
```
[2/4] Fetching packages...
info fsevents@2.1.2: The platform "linux" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.3.2: The platform "linux" is incompatible with this module.
info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.9: The platform "linux" is incompatible with this module.
info "fsevents@1.2.9" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > boxwise-components@1.0.4" has unmet peer dependency "rebass@4.x".
warning " > boxwise-components@1.0.4" has unmet peer dependency "theme-ui@0.3x".
warning "boxwise-components > @rollup/plugin-json@4.1.0" has unmet peer dependency "rollup@^1.20.0 || ^2.0.0".
warning "boxwise-components > rollup-plugin-node-resolve@5.2.0" has unmet peer dependency "rollup@>=1.11.0".
warning "boxwise-components > rollup-plugin-terser@7.0.2" has unmet peer dependency "rollup@^2.0.0".
warning "boxwise-components > @rollup/plugin-json > @rollup/pluginutils@3.1.0" has unmet peer dependency "rollup@^1.20.0||^2.0.0".
warning "boxwise-components > theme-ui > @theme-ui/mdx > @mdx-js/react@1.6.22" has incorrect peer dependency "react@^16.13.1 || ^17.0.0".
warning " > @semantic-ui-react/craco-less@1.2.1" has incorrect peer dependency "react-scripts@^3.1.0".
warning "@semantic-ui-react/craco-less > craco-less@1.17.0" has incorrect peer dependency "react-scripts@^3.3.0".
warning "@semantic-ui-react/craco-less > craco-less > less-loader@6.2.0" has unmet peer dependency "webpack@^4.0.0 || ^5.0.0".
warning " > @testing-library/user-event@13.2.1" has unmet peer dependency "@testing-library/dom@>=7.21.4".
warning " > eslint-config-airbnb@18.1.0" has unmet peer dependency "eslint@^5.16.0 || ^6.8.0".
warning " > eslint-config-airbnb@18.1.0" has unmet peer dependency "eslint-plugin-import@^2.20.1".
warning " > eslint-config-airbnb@18.1.0" has unmet peer dependency "eslint-plugin-react@^7.19.0".
warning " > eslint-config-airbnb@18.1.0" has unmet peer dependency "eslint-plugin-react-hooks@^2.5.0 || ^1.7.0".
warning "eslint-config-airbnb > eslint-config-airbnb-base@14.1.0" has unmet peer dependency "eslint@^5.16.0 || ^6.8.0".
warning "eslint-config-airbnb > eslint-config-airbnb-base@14.1.0" has unmet peer dependency "eslint-plugin-import@^2.20.1".
warning " > eslint-config-prettier@6.11.0" has unmet peer dependency "eslint@>=3.14.1".
warning " > eslint-plugin-jsx-a11y@6.2.3" has unmet peer dependency "eslint@^3 || ^4 || ^5 || ^6".
warning "eslint-plugin-jsx-a11y > axobject-query@2.1.1" has unmet peer dependency "eslint@^5 || ^6".
warning " > eslint-plugin-prettier@3.1.3" has unmet peer dependency "eslint@>= 5.0.0".
```
and when I start the container and run `yarn list` I can see the the `auth0-react` package is installed.

aerinsol commented:
Regarding testing, I'm happy for it to be more lax as we have no user
facing production code on this repo right now, but I WOULD like us to move
to a more strict TDD methodology soon as that changes.

On Tue, Sep 21, 2021, 4:17 PM P M ***@***.***> wrote:

> Hej Daniel, all good! I moved a couple of messages to Slack to keep the
> postings here about PR-related stuff. I won't push to this branch anymore =)
>
>    - For frontend coverage, I don't really have an opinion. For BE
>    coverage I try to be strict and to cover whatever new production code is
>    implemented
>    - rebasing the current branch on master could become tricky now that
>    #59 <https://github.com/boxwise/boxtribute/pull/59> was merged...maybe
>    now it is a good idea to remove the resolver implementations and their
>    tests again
>    - I still can't run the react service. With docker-compose up react I
>    have the same error about auth0-react missing. When I run docker build
>    react/ before I receive
>
> [2/4] Fetching packages...
> info ***@***.***: The platform "linux" is incompatible with this module.
> info ***@***.***" is an optional dependency and failed compatibility check. Excluding it from installation.
> info ***@***.***: The platform "linux" is incompatible with this module.
> info ***@***.***" is an optional dependency and failed compatibility check. Excluding it from installation.
> info ***@***.***: The platform "linux" is incompatible with this module.
> info ***@***.***" is an optional dependency and failed compatibility check. Excluding it from installation.
> [3/4] Linking dependencies...
> warning " > ***@***.***" has unmet peer dependency ***@***.***".
> warning " > ***@***.***" has unmet peer dependency ***@***.***".
> warning "boxwise-components > @***@***.***" has unmet peer dependency "rollup@^1.20.0 || ^2.0.0".
> warning "boxwise-components > ***@***.***" has unmet peer dependency "rollup@>=1.11.0".
> warning "boxwise-components > ***@***.***" has unmet peer dependency "rollup@^2.0.0".
> warning "boxwise-components > @rollup/plugin-json > @***@***.***" has unmet peer dependency "rollup@^1.20.0||^2.0.0".
> warning "boxwise-components > theme-ui > @theme-ui/mdx > @***@***.***" has incorrect peer dependency "react@^16.13.1 || ^17.0.0".
> warning " > @***@***.***" has incorrect peer dependency "react-scripts@^3.1.0".
> warning ***@***.***/craco-less > ***@***.***" has incorrect peer dependency "react-scripts@^3.3.0".
> warning ***@***.***/craco-less > craco-less > ***@***.***" has unmet peer dependency "webpack@^4.0.0 || ^5.0.0".
> warning " > @***@***.***" has unmet peer dependency ***@***.***/dom@>=7.21.4".
> warning " > ***@***.***" has unmet peer dependency "eslint@^5.16.0 || ^6.8.0".
> warning " > ***@***.***" has unmet peer dependency "eslint-plugin-import@^2.20.1".
> warning " > ***@***.***" has unmet peer dependency "eslint-plugin-react@^7.19.0".
> warning " > ***@***.***" has unmet peer dependency "eslint-plugin-react-hooks@^2.5.0 || ^1.7.0".
> warning "eslint-config-airbnb > ***@***.***" has unmet peer dependency "eslint@^5.16.0 || ^6.8.0".
> warning "eslint-config-airbnb > ***@***.***" has unmet peer dependency "eslint-plugin-import@^2.20.1".
> warning " > ***@***.***" has unmet peer dependency "eslint@>=3.14.1".
> warning " > ***@***.***" has unmet peer dependency "eslint@^3 || ^4 || ^5 || ^6".
> warning "eslint-plugin-jsx-a11y > ***@***.***" has unmet peer dependency "eslint@^5 || ^6".
> warning " > ***@***.***" has unmet peer dependency "eslint@>= 5.0.0".
>
> and when I start the container and run yarn list I can see the the
> auth0-react package is installed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/boxwise/boxtribute/pull/62#issuecomment-923748671>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABP2BFAZWWS7TEUQUL2NGE3UDA5TJANCNFSM5EJNC6ZA>
> .
>

spaudanjo commented:
@pylipp 
`I still can't run the react service. With docker-compose up react I have the same error about auth0-react missing.` => That's pretty weird. I'm able to run `docker-compose up` and can open the react app (the branch/commit pushed here). 
It's weird to say since it's exactly the thing that docker aims to overcome, but I have to say `Works on my machine` :-/ :D 
Even though it shouldn't make a difference, I can as a next action remove my docker images and build them again from scratch.

spaudanjo commented:
Just deleted my `node_modules`, deleted the react docker container and image and started ran `docker-compose up` again - everything worked (besides the - expected - fact that the FE is not fully functional anymore since the queries are now out of sync with the schema changes of course :D but as discussed, I can do this in a next PR)! 
@vahidbazzaz could you maybe try to run this branch via docker-compose and see whether you can open the frontend?

spaudanjo commented:
and since @aerinsol and @pylipp agree with a temporary decrease of react app codecov (+1 regarding @aerinsol TDD comment - latest with the discussed overall review of the whole current FE app I think this would make sense): could someone adapt it accordingly so we can merge this PR? I think I don't the rights to do it.

vahidbazzaz commented:
@spaudanjo sure, I will run and back to you, so we can figure it out hopefully.

vahidbazzaz commented:
@spaudanjo I have removed my node_module and increased the RAM and Docker timeout, now your branch is working fine at my end.

pylipp commented:
Thanks for your advices :D I have completely removed all services and images and rebuilt everything from scratch. Now I also see the memory error (JavaScript heap out of memory). In my (Linux) Docker installation there is no UI to set the memory limits globally. I had to specify `version: "2.4"` and then use the `mem_limit: 4G` option in `docker-compose.yml`. I still can't start the react service. Now I notice that my hardware is not powerful enough (`free -th` shows 3.6G free memory). So I think it would work overall, and I'll merge this PR!

pylipp reviewed (approved):
Good thing to have cleared out all changes in the `flask/` directory.
I hope the changes in #67 are not too confusing from a frontend POV.

spaudanjo commented:
Amazing! :D
